File: app/src/main/java/com/agrogames/islandsofwar/engine/abs/transport/TransportUnit.java
package com.agrogames.islandsofwar.engine.abs.transport;

import android.os.Build;

import androidx.annotation.RequiresApi;

import com.agrogames.islandsofwar.engine.abs.common.Cell;
import com.agrogames.islandsofwar.engine.abs.unit.IUnit;

import java.util.Objects;
import java.util.concurrent.Callable;
import java.util.function.Function;

public class TransportUnit {
    public final Function<Cell, IUnit> create;
    public final IUnit example;

    @RequiresApi(api = Build.VERSION_CODES.N)
    public TransportUnit(Function<Cell, IUnit> create) {
        this.create = create;
        example = create.apply(new Cell(-1, -1));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TransportUnit that = (TransportUnit) o;
        return Objects.equals(example.getTexture(), that.example.getTexture());
    }

    @Override
    public int hashCode() {
        return Objects.hash(example.getTexture());
    }
}

File: app/src/main/java/com/agrogames/islandsofwar/engine/abs/transport/TransportUnit.kt
package com.agrogames.islandsofwar.engine.abs.transport

import com.agrogames.islandsofwar.engine.abs.common.Cell
import com.agrogames.islandsofwar.engine.abs.unit.IUnit
import java.util.Objects
import java.util.function.Function

class TransportUnit constructor(val create: Function<Cell, IUnit>) {
    val example = create.apply(Cell(-1, -1))

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        return example.texture == (other as TransportUnit).example.texture
    }

    override fun hashCode(): Int {
        return Objects.hash(example.texture)
    }
}